### PR TITLE
[MI-2343][server+webapp]: Added support to show pipeline release link preview.

### PR DESCRIPTION
**Pull request by avas27JTG**
Added support to show pipeline release link preview
Made changes to show the pipeline icon completely in MM post

image

### CODE REVIEW

File: server/plugin/api.go [hunk 1]
```diff
@@ -417,6 +417,21 @@ func (p *Plugin) getReviewersListString(reviewersList []serializers.Reviewer) st
 	return reviewers
 }
 
+func (p *Plugin) getPipelineReleaseEnvironmentList(environments []serializers.Environments) string {
```
Comment (ayusht2810) [inline server/plugin/api.go:420]:
```suggestion
func (p *Plugin) getPipelineReleaseEnvironmentList(environments []*serializers.Environments) string {
```

File: server/plugin/api.go [hunk 2]
```diff
@@ -417,6 +417,21 @@ func (p *Plugin) getReviewersListString(reviewersList []serializers.Reviewer) st
...
+		envs += env.Name
+		if count != (len(environments) - 1) {
+			envs += " | "
+		}
+	}
+
+	if envs == "" {
+		return "None"
+	}
+	return envs
```
Comment (ayusht2810) [inline server/plugin/api.go:432]:
```suggestion
	}

	return envs
```

File: server/plugin/taskLinkPreview.go [hunk 3]
```diff
@@ -154,3 +154,42 @@ func (p *Plugin) PostBuildDetailsPreview(linkData []string, link, userID, channe
...
+	releaseDetails, _, err := p.Client.GetReleaseDetails(organization, project, releaseID, userID)
+	if err != nil {
+		return nil, ""
+	}
+
+	post := &model.Post{
+		UserId:    userID,
+		ChannelId: channelID,
+	}
+	environments := p.getPipelineReleaseEnvironmentList(releaseDetails.Environments)
```
Comment (ayusht2810) [inline server/plugin/taskLinkPreview.go:171]:
```suggestion
	}
	
	environments := p.getPipelineReleaseEnvironmentList(releaseDetails.Environments)
```

File: server/plugin/taskLinkPreview.go [hunk 4]
```diff
@@ -154,3 +154,42 @@ func (p *Plugin) PostBuildDetailsPreview(linkData []string, link, userID, channe
...
+				Title: "Status",
+				Value: releaseDetails.Status,
+				Short: true,
+			},
+		},
+		Footer:     project,
+		FooterIcon: fmt.Sprintf("%s/plugins/%s/static/%s", p.GetSiteURL(), constants.PluginID, constants.FileNameProjectIcon),
+	}
+	model.ParseSlackAttachment(post, []*model.SlackAttachment{attachment})
+
```
Comment (ayusht2810) [inline server/plugin/taskLinkPreview.go:193]:
```suggestion
	}

	model.ParseSlackAttachment(post, []*model.SlackAttachment{attachment})
```

File: server/serializers/subscriptions.go [hunk 5]
```diff
@@ -176,6 +176,23 @@ type Definition struct {
...
+	ReleaseDefinition ReleaseDefinition `json:"releaseDefinition"`
+}
+
+type Environments struct {
+	Name string `json:"name"`
+}
+
+type ReleaseDefinition struct {
+	Name string `json:"name"`
+}
```
Comment (ayusht2810) [inline server/serializers/subscriptions.go:194]: https://github.com/Brightscout/mattermost-plugin-azure-devops/pull/104#discussion_r1018774655
same

File: server/plugin/client.go [hunk 6]
```diff
@@ -130,6 +131,21 @@ func (c *client) GetBuildDetails(organization, projectName, buildID, mattermostU
 	return buildDetails, statusCode, nil
 }
 
+// Function to get the pipeline release details.
+func (c *client) GetReleaseDetails(organization, projectName, releaseID, mattermostUserID string) (*serializers.ReleaseDetails, int, error) {
+	buildDetailsURL := fmt.Sprintf(constants.GetReleaseDetails, organization, projectName, releaseID)
```
Comment (raghavaggarwal2308) [inline server/plugin/client.go:136]:
```suggestion
	releaseDetailsURL := fmt.Sprintf(constants.GetReleaseDetails, organization, projectName, releaseID)
```

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin/taskLinkPreview.go (hunk 3 above)
Comment (avas27JTG) [reply server/plugin/taskLinkPreview.go:175]: can you let me know on what basis you decide to add a line break?

File: server/serializers/subscriptions.go (hunk 5 above)
Comment (avas27JTG) [reply server/serializers/subscriptions.go:264]: here although we have other structs with `name` JSON field but created this separate for code readability and tally with API response as using a generic one may create confusion to match with API response structure

File: server/plugin/taskLinkPreview.go (hunk 3 above)
Comment (ayusht2810) [reply server/plugin/taskLinkPreview.go]: I usually add a line break at the end of a block

File: server/plugin/api.go [hunk 7]
```diff
@@ -220,6 +229,26 @@ func (p *Plugin) handleUnlinkProject(w http.ResponseWriter, r *http.Request) {
...
+		if subscription.ProjectID == projectID {
+			statusCode, deleteErr := p.deleteSubscription(subscription, mattermostUserID)
+			if deleteErr != nil {
+				p.API.LogError(constants.DeleteSubscriptionError, "Error", deleteErr.Error())
+				return statusCode, deleteErr
+			}
+		}
+	}
+
+	return http.StatusOK, nil
```
Comment (ayusht2810) [inline server/plugin/api.go:249]: can't we use a function `returnStatusOK` here which is already present?

File: server/plugin/api.go [hunk 8]
```diff
@@ -554,6 +583,22 @@ func (p *Plugin) handleSubscriptionNotifications(w http.ResponseWriter, r *http.
...
+		// Convert map to json string
+		jsonBytes, err := json.Marshal(body.Resource.Comment)
+		if err != nil {
+			p.API.LogError(err.Error())
+			p.handleError(w, r, &serializers.Error{Code: http.StatusInternalServerError, Message: err.Error()})
+			return
+		}
+
+		// Convert json string to struct
+		var comment serializers.Comment
```
Comment (ayusht2810) [inline server/plugin/api.go:595]:
```suggestion
		var comment *serializers.Comment
```

File: server/plugin/api.go [hunk 9]
```diff
@@ -660,32 +904,26 @@ func (p *Plugin) handleDeleteSubscriptions(w http.ResponseWriter, r *http.Reques
...
+}
+
+func (p *Plugin) deleteSubscription(subscription *serializers.SubscriptionDetails, mattermostUserID string) (int, error) {
+	if statusCode, err := p.Client.DeleteSubscription(subscription.OrganizationName, subscription.SubscriptionID, mattermostUserID); err != nil {
+		return statusCode, err
 	}
 
-	returnStatusOK(w)
+	if deleteErr := p.Store.DeleteSubscription(subscription); deleteErr != nil {
+		return http.StatusInternalServerError, deleteErr
```
Comment (ayusht2810) [inline server/plugin/api.go:923]: same

File: server/plugin/api.go [hunk 10]
```diff
@@ -660,32 +904,26 @@ func (p *Plugin) handleDeleteSubscriptions(w http.ResponseWriter, r *http.Reques
...
+	}
+
+	return http.StatusOK, nil
```
Comment (ayusht2810) [inline server/plugin/api.go:926]: same

File: server/plugin/api_test.go [hunk 11]
```diff
@@ -279,7 +279,8 @@ func TestHandleLink(t *testing.T) {
 				mockedStore.EXPECT().GetAllProjects("mockMattermostUserID").Return(testCase.projectList, nil)
 				mockedStore.EXPECT().StoreProject(&serializers.ProjectDetails{
 					MattermostUserID: "mockMattermostUserID",
-					OrganizationName: "mockOrganization",
+					ProjectName:      "Mockproject",
+					OrganizationName: "mockorganization",
```
Comment (ayusht2810) [inline server/plugin/api_test.go:283]:
```suggestion
					ProjectName:      "mockProject",
					OrganizationName: "mockOrganization",
```

Comment (raghavaggarwal2308) [reply server/plugin/api_test.go]: We are making the project name tilte case and org name as smallcase here

File: server/plugin/api.go [hunk 12]
```diff
@@ -220,6 +229,26 @@ func (p *Plugin) handleUnlinkProject(w http.ResponseWriter, r *http.Request) {
 	p.writeJSON(w, &successResponse)
 }
 
+func (p *Plugin) handleDeleteAllSubscriptions(mattermostUserID, projectID string) (int, error) {
+	subscriptionList, err := p.Store.GetAllSubscriptions(mattermostUserID)
+	if err != nil {
+		p.API.LogError(constants.FetchSubscriptionListError, "Error", err.Error())
+		return http.StatusInternalServerError, err
```
[the remark this answers is not in the document]
Comment (ayusht2810) [reply server/plugin/api.go]: then shouldn't we move this to some other file like util.go?
also, I don't see any problem using `p.handleError`

File: server/plugin/api.go [hunk 13]
```diff
@@ -488,6 +488,22 @@ func (p *Plugin) getReviewersListString(reviewersList []serializers.Reviewer) st
 	return reviewers
 }
 
+func (p *Plugin) getPipelineReleaseEnvironmentList(environments []*serializers.Environments) string {
+	envs := ""
+	for count, env := range environments {
```
Comment (manojmalik20) [inline server/plugin/api.go:493]:
```suggestion
	for index, env := range environments {
```

File: server/plugin/taskLinkPreview.go [hunk 14]
```diff
@@ -159,3 +159,44 @@ func (p *Plugin) PostBuildDetailsPreview(linkData []string, link, userID, channe
 	model.ParseSlackAttachment(post, []*model.SlackAttachment{attachment})
 	return post, ""
 }
+
+func (p *Plugin) PostReleaseDetailsPreview(linkData []string, link, userID, channelID string) (*model.Post, string) {
+	organization := linkData[3]
+	project := linkData[4]
+	releaseID := strings.Split(linkData[5], "&")[1][10:]
+	releaseDetails, _, err := p.Client.GetReleaseDetails(organization, project, releaseID, userID)
+	if err != nil {
```
Comment (manojmalik20) [inline server/plugin/taskLinkPreview.go:168]:
```suggestion
	if err != nil {
	p.API.LogDebug(.....
```
I think we should log the error with debug level.

File: server/plugin/taskLinkPreview.go [hunk 15]
```diff
@@ -159,3 +159,44 @@ func (p *Plugin) PostBuildDetailsPreview(linkData []string, link, userID, channe
...
+	}
+
+	post := &model.Post{
+		UserId:    userID,
+		ChannelId: channelID,
+	}
+
+	environments := p.getPipelineReleaseEnvironmentList(releaseDetails.Environments)
+	attachment := &model.SlackAttachment{
+		AuthorName: "Azure Pipeline",
```
Comment (manojmalik20) [inline server/plugin/taskLinkPreview.go:179]:
```suggestion
		AuthorName: "Azure Pipelines",
```
Shouldn't it be this?

File: server/plugin/taskLinkPreview.go [hunk 16]
```diff
@@ -159,3 +159,44 @@ func (p *Plugin) PostBuildDetailsPreview(linkData []string, link, userID, channe
...
+		AuthorIcon: fmt.Sprintf("%s/plugins/%s/static/%s", p.GetSiteURL(), constants.PluginID, constants.FileNamePipelinesIcon),
+		Title:      fmt.Sprintf(constants.PipelineDetailsTitle, releaseDetails.Name, releaseDetails.Link.Web.Href, releaseDetails.ReleaseDefinition.Name),
```
Comment (manojmalik20) [inline server/plugin/taskLinkPreview.go:181]: Why aren't we using the `TitleLink` property of the slack attachment?

File: server/plugin/taskLinkPreview.go [hunk 17]
```diff
@@ -159,3 +159,44 @@ func (p *Plugin) PostBuildDetailsPreview(linkData []string, link, userID, channe
...
+
+	post := &model.Post{
+		UserId:    userID,
+		ChannelId: channelID,
+	}
+
+	environments := p.getPipelineReleaseEnvironmentList(releaseDetails.Environments)
+	attachment := &model.SlackAttachment{
+		AuthorName: "Azure Pipeline",
+		AuthorIcon: fmt.Sprintf("%s/plugins/%s/static/%s", p.GetSiteURL(), constants.PluginID, constants.FileNamePipelinesIcon),
```
Comment (manojmalik20) [inline server/plugin/taskLinkPreview.go:180]: I think we should use the `public` folder for hosting all the icons and files.

File: server/plugin/taskLinkPreview.go [hunk 18]
```diff
@@ -159,3 +159,44 @@ func (p *Plugin) PostBuildDetailsPreview(linkData []string, link, userID, channe
...
+				Value: releaseDetails.Status,
+				Short: true,
+			},
+		},
+		Footer:     project,
+		FooterIcon: fmt.Sprintf("%s/plugins/%s/static/%s", p.GetSiteURL(), constants.PluginID, constants.FileNameProjectIcon),
+	}
+
+	model.ParseSlackAttachment(post, []*model.SlackAttachment{attachment})
+
```
Comment (manojmalik20) [inline server/plugin/taskLinkPreview.go:200]:
```suggestion
```

File: server/serializers/subscriptions.go [hunk 19]
```diff
@@ -284,6 +284,23 @@ type RequestedBy struct {
...
+type ReleaseDetails struct {
+	Name              string            `json:"name"`
+	ID                int               `json:"id"`
+	Status            string            `json:"status"`
+	Environments      []*Environments   `json:"environments"`
+	Link              Link              `json:"_links"`
+	ReleaseDefinition ReleaseDefinition `json:"releaseDefinition"`
+}
+
+type Environments struct {
```
Comment (manojmalik20) [inline server/serializers/subscriptions.go:296]:
```suggestion
type Environment struct {
```

File: server/serializers/subscriptions.go [hunk 20]
```diff
@@ -284,6 +284,23 @@ type RequestedBy struct {
 	DisplayName string `json:"displayName"`
 }
 
+type ReleaseDetails struct {
+	Name              string            `json:"name"`
+	ID                int               `json:"id"`
+	Status            string            `json:"status"`
+	Environments      []*Environments   `json:"environments"`
```
Comment (manojmalik20) [inline server/serializers/subscriptions.go:291]:
```suggestion
	Environments      []*Environment   `json:"environments"`
```

File: server/plugin/taskLinkPreview.go (hunk 17 above)
Comment (avas27JTG) [reply server/plugin/taskLinkPreview.go:181]: Yes, have already logged this here
https://github.com/mattermost/mattermost-plugin-azure-devops/pull/104#discussion_r1048073930

File: server/plugin/taskLinkPreview.go (hunk 16 above)
Comment (avas27JTG) [reply server/plugin/taskLinkPreview.go:182]: I was not aware of it and in many places this was already used by other devs, taking a note we can do it after merging everything.

**Review by ayusht2810 — APPROVED** [review]

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**